### PR TITLE
Add hooks to Mercurius

### PR DESCRIPTION
**Pull request by projectjudge**
Closes: https://github.com/mercurius-js/mercurius/issues/380

### CODE REVIEW

File: lib/hooks.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+      }
+    }
+    await next(null, newDocument, newErrors)
+  }
+
+  function handleReject (err) {
+    cb(err, request, executionResult)
+  }
+
+  await next()
```
Comment (mcollina) [inline lib/hooks.js:137]: Could you please avoid the recursion and use a loop instead? We are just using promises here, there is no need to have a callback.

File: lib/hooks.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  }
+
+  function handleReject (err) {
+    cb(err, request)
+  }
+
+  await next()
+}
+
+async function preExecutionHookRunner (functions, request, executionResult, cb) {
```
Comment (mcollina) [inline lib/hooks.js:77]: don't mix callbacks and async

File: lib/hooks.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+  async function handleResolve () {
+    await next()
+  }
+
+  function handleReject (err) {
+    cb(err, request)
+  }
+
+  await next()
```
Comment (mcollina) [inline lib/hooks.js:74]: use a loop, don't recurse with async.

File: test/gateway/hooks.js [hunk 4]
```diff
@@ -0,0 +1,1075 @@
+'use strict'
+
+const { test } = require('tap')
+const Fastify = require('fastify')
+const { GraphQLSchema, parse } = require('graphql')
+const GQL = require('../..')
+
+const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms))
```
Comment (mcollina) [inline test/gateway/hooks.js:8]: Please use `promisify(setTimeout)` instead

**Review by mcollina — CHANGES_REQUESTED** [review]
I've done a quick pass, I need to get deeper but I think there are a few things to fix already.

File: lib/hooks.js (hunk 1 above)
Comment (projectjudge) [reply lib/hooks.js:137]: yep, no problem

File: lib/handlers.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+const { print } = require('graphql')
+
+async function preParsingHandler (request) {
+  if (request.context.preParsing !== null) {
+    await hooksRunner(
+      request.context.preParsing,
+      preParsingHookRunner,
+      request
+    )
+  }
```
Comment (mcollina) [inline lib/handlers.js:12]: The if should be moved to the caller instead.

File: lib/handlers.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+async function preValidationHandler (request) {
+  if (request.context.preValidation !== null) {
+    await hooksRunner(
+      request.context.preValidation,
+      hookRunner,
+      request
+    )
+  }
```
Comment (mcollina) [inline lib/handlers.js:22]: The if should be moved to the caller instead.

File: lib/handlers.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      request.context.preExecution,
+      request
+    )
+    if (errors.length > 0) {
+      addErrorsToContext(request.context, errors)
+    }
+    if (typeof modifiedDocument !== 'undefined') {
+      return { modifiedDocument, modifiedQuery: print(request.document) }
+    }
+  }
```
Comment (mcollina) [inline lib/handlers.js:37]: The if should be moved to the caller instead.

File: lib/handlers.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      request.context.preGatewayExecution,
+      request
+    )
+    if (errors.length > 0) {
+      addErrorsToContext(request.context, errors)
+    }
+    if (typeof modifiedDocument !== 'undefined') {
+      return { modifiedDocument, modifiedQuery: print(modifiedDocument) }
+    }
+  }
```
Comment (mcollina) [inline lib/handlers.js:53]: The if should be moved to the caller instead.

File: lib/handlers.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+async function onResolutionHandler (request) {
+  if (request.context.onResolution !== null) {
+    await hooksRunner(
+      request.context.onResolution,
+      onResolutionHookRunner,
+      request
+    )
+  }
```
Comment (mcollina) [inline lib/handlers.js:64]: The if should be moved to the caller instead.

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work! How do the hooks play with subscriptions?

**Comment by projectjudge** [comment]
Thanks! I have tested on a very basic level and seems to be playing okay. I'll write some thorough tests for subscriptions for both normal server and gateway mode to make sure it is all working as expected :) I also noticed that the CI checks are intermittently failing for some of the subscription tests. At first glance it seems to be an environmental issue (other checks for the same node version pass fine) - did you have any thoughts on this? (Example run: https://github.com/mercurius-js/mercurius/pull/385/checks?check_run_id=1791951386)

**Comment by mcollina** [comment]
I've seen the flaky tests... can't do much about it right now. If you'd like to investigate and fix them, I'd be very grateful!

**Comment by projectjudge** [comment]
For sure, very happy to take a look at the tests! :)

File: lib/hooks.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  this.preValidation = []
+  this.preExecution = []
+  this.preGatewayExecution = []
+  this.onResolution = []
+}
+
+Hooks.prototype.validate = function (hook, fn) {
+  if (typeof hook !== 'string') throw new MER_ERR_HOOK_INVALID_TYPE()
+  if (typeof fn !== 'function') throw new MER_ERR_HOOK_INVALID_HANDLER()
+  if (supportedHooks.indexOf(hook) === -1) {
```
Comment (mcollina) [inline lib/hooks.js:18]:
```suggestion
  if (lifecycleHooks.indexOf(hook) === -1) {
```

I think you can remove the `supportedHooks` variable

File: test/hooks.js [hunk 11]
```diff
@@ -0,0 +1,682 @@
+'use strict'
+
+const { test } = require('tap')
+const Fastify = require('fastify')
+const { parse, GraphQLSchema } = require('graphql')
+const GQL = require('..')
+
+const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms))
```
Comment (mcollina) [inline test/hooks.js:8]:
```suggestion
const { promisify } = require('util')
const sleep = promisify(setTimeout)
```

File: test/hooks.js [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+
+// -----
+// hooks
+// -----
+test('hooks', async t => {
+  t.plan(16)
+  const app = await createTestServer(t)
+
+  app.graphql.addHook('preParsing', async function (schema, source, context) {
+    await sleep(1)
```
Comment (mcollina) [inline test/hooks.js:51]: why is this needed?

File: test/hooks.js [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+    }
+  }
+}
+
+const query = '{ add(x: 2, y: 2) }'
+
+async function createTestServer (t, customResolvers = resolvers) {
+  const app = Fastify()
+  t.tearDown(() => {
+    app.close()
```
Comment (mcollina) [inline test/hooks.js:32]:
```suggestion
    return app.close()
```

File: test/hooks.js [hunk 14]
```diff
@@ -0,0 +1,682 @@
...
+  })
+
+  app.register(GQL, { schema, resolvers: customResolvers })
+
+  // needed so that graphql is defined
+  await app.ready()
```
Comment (mcollina) [inline test/hooks.js:38]: This is actually not needed anymore, just returning app would do.

File: test/hooks.js [hunk 15]
```diff
@@ -0,0 +1,682 @@
...
+
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.fail('this should not be called')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:203]: this is not needed at all.

File: test/hooks.js [hunk 16]
```diff
@@ -0,0 +1,682 @@
...
+  })
+
+  app.graphql.addHook('preParsing', async (schema, source, context) => {
+    t.type(schema, GraphQLSchema)
+    t.is(source, query)
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:240]: not needed

File: test/hooks.js [hunk 17]
```diff
@@ -0,0 +1,682 @@
...
+
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.fail('this should not be called')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:282]: not needed

File: test/hooks.js [hunk 18]
```diff
@@ -0,0 +1,682 @@
...
+  })
+
+  app.graphql.addHook('preValidation', async (schema, document, context) => {
+    t.type(schema, GraphQLSchema)
+    t.deepEqual(document, parse(query))
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:319]: not needed

File: test/hooks.js [hunk 19]
```diff
@@ -0,0 +1,682 @@
...
+
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.fail('this should not be called')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:357]: not needed

File: test/hooks.js [hunk 20]
```diff
@@ -0,0 +1,682 @@
...
+  })
+
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.type(schema, GraphQLSchema)
+    t.deepEqual(document, parse(query))
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:394]: not needed

File: test/hooks.js [hunk 21]
```diff
@@ -0,0 +1,682 @@
...
+        }
+      ],
+      loc: { start: 0, end: 19 }
+    }
+    return {
+      document: modifiedDocument
+    }
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:464]: not needed

File: test/hooks.js [hunk 22]
```diff
@@ -0,0 +1,682 @@
...
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.type(schema, GraphQLSchema)
+    t.deepEqual(document, parse(query))
+    t.type(context, 'object')
+    return {
+      errors: [new Error('bar')]
+    }
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:504]: not needed

File: test/hooks.js [hunk 23]
```diff
@@ -0,0 +1,682 @@
...
```
Comment (mcollina) [inline test/hooks.js:560]: not needed

File: test/hooks.js [hunk 24]
```diff
@@ -0,0 +1,682 @@
...
+// -------------
+test('preGatewayExecution hooks should never be called in a non-gateway graphql server', async t => {
+  t.plan(1)
+  const app = await createTestServer(t)
+
+  app.graphql.addHook('preGatewayExecution', async (schema, document, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:600]: not needed

File: test/hooks.js [hunk 25]
```diff
@@ -0,0 +1,682 @@
...
+    t.type(execution, 'object')
+    t.type(context, 'object')
+    throw new Error('a onResolution error occured')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:633]: not needed

File: test/hooks.js [hunk 26]
```diff
@@ -0,0 +1,682 @@
...
+    context.foo = 'bar'
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.type(execution, 'object')
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:668]: not needed

File: test/gateway/hooks.js [hunk 27]
```diff
@@ -0,0 +1,1076 @@
...
+      services: [{
+        name: 'user',
+        url: `http://localhost:${userServicePort}/graphql`
+      }, {
+        name: 'post',
+        url: `http://localhost:${postServicePort}/graphql`
+      }]
+    }
+  })
+  await gateway.ready()
```
Comment (mcollina) [inline test/gateway/hooks.js:160]: not needed

File: test/gateway/hooks.js [hunk 28]
```diff
@@ -0,0 +1,1076 @@
...
+  })
+
+  // This should still be called in the gateway
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.type(execution, 'object')
+    t.type(context, 'object')
+    t.ok('onResolution called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:751]: not needed

File: test/gateway/hooks.js [hunk 29]
```diff
@@ -0,0 +1,1076 @@
...
+    t.type(context, 'object')
+    t.ok('preExecution called')
+    const documentClone = JSON.parse(JSON.stringify(document))
+    documentClone.definitions[0].selectionSet.selections = [documentClone.definitions[0].selectionSet.selections[0]]
+    return {
+      document: documentClone
+    }
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:620]: not needed

File: test/gateway/hooks.js [hunk 30]
```diff
@@ -0,0 +1,1076 @@
...
+  })
+
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.type(schema, GraphQLSchema)
+    t.deepEqual(document, parse(query))
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:563]: not  neeeded

File: test/gateway/hooks.js [hunk 31]
```diff
@@ -0,0 +1,1076 @@
...
+
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.fail('this should not be called')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:526]: not needed

File: test/gateway/hooks.js [hunk 32]
```diff
@@ -0,0 +1,1076 @@
...
+  app.graphql.addHook('preExecution', async (schema, document, context) => {
+    t.fail('this should not be called')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    await sleep(1)
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:426]: not needed

File: test/gateway/hooks.js [hunk 33]
```diff
@@ -0,0 +1,1076 @@
...
+  })
+
+  app.graphql.addHook('preParsing', async (schema, source, context) => {
+    t.type(schema, GraphQLSchema)
+    t.is(source, query)
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:358]: not needed

File: test/gateway/hooks.js [hunk 34]
```diff
@@ -0,0 +1,1076 @@
...
+
+  app.graphql.addHook('preExecution', async (schema, operation, context) => {
+    t.fail('this should not be called')
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.fail('this should not be called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:321]: not needed

File: test/gateway/hooks.js [hunk 35]
```diff
@@ -0,0 +1,1076 @@
...
+  })
+
+  app.graphql.addHook('preValidation', async (schema, document, context) => {
+    t.type(schema, GraphQLSchema)
+    t.deepEqual(document, parse(query))
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:463]: not needed

File: test/gateway/hooks.js [hunk 36]
```diff
@@ -0,0 +1,1076 @@
...
+    t.type(schema, GraphQLSchema)
+    t.type(document, 'object')
+    t.type(context, 'object')
+    t.ok('preGatewayExecution called for foo error')
+    return {
+      errors: [new Error(`bar - ${document.definitions[0].name.value}`)]
+    }
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:863]: not needed

File: test/gateway/hooks.js [hunk 37]
```diff
@@ -0,0 +1,1076 @@
...
+    context.foo = 'bar'
+  })
+
+  app.graphql.addHook('onResolution', async (execution, context) => {
+    t.type(execution, 'object')
+    t.type(context, 'object')
+    t.is(context.foo, 'bar')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/gateway/hooks.js:1037]: not needed

File: test/hooks.js [hunk 38]
```diff
@@ -0,0 +1,682 @@
...
+  })
+
+  app.graphql.addHook('onResolution', async function (execution, context) {
+    await sleep(1)
+    t.type(execution, 'object')
+    t.type(context, 'object')
+    t.ok('onResolution called')
+  })
+
+  await app.listen(0)
```
Comment (mcollina) [inline test/hooks.js:86]: not needed

File: lib/gateway/make-resolver.js [hunk 39]
```diff
@@ -454,7 +463,8 @@ function makeResolver ({ service, createOperation, transformData, isQuery, isRef
     return reply[kEntityResolvers][`${service.name}Entity`]({
       query,
       variables,
-      originalRequestHeaders: reply.request.headers
+      originalRequestHeaders: reply.request.headers,
+      context
     }).then(transformData)
```
Comment (mcollina) [inline lib/gateway/make-resolver.js:468]: please change then to `await`, `makeResolver` is now an async function (also a few lines above)

**Review by mcollina — COMMENTED** [review]
Good work! I did another pass and I've left a few comments.

File: test/hooks.js (hunk 12 above)
Comment (projectjudge) [reply test/hooks.js:51]: I wanted to have an explicit promise to await within the hooks function so we can simulate some work - similar to how fastify tests hooks. I'll remove it however, because it's not really needed for these test cases and what is currently there in the function definitions should be sufficient!

File: test/hooks.js (hunk 14 above)
Comment (projectjudge) [reply test/hooks.js:38]: Nice, that's good to know!

File: lib/gateway/make-resolver.js (hunk 39 above)
Comment (projectjudge) [reply lib/gateway/make-resolver.js:468]: sure!

File: test/hooks.js (hunk 12 above)
Comment (mcollina) [reply test/hooks.js:51]: You can do:

```js
const immediate = promisify(setImmediate).
```

Comment (projectjudge) [reply test/hooks.js:51]: ah nice - thanks! I'll push up the rest of these changes you've suggested now and get to work on adding thorough tests for subscriptions

File: lib/handlers.js [hunk 40]
```diff
@@ -0,0 +1,57 @@
+const { hooksRunner, preExecutionHooksRunner, hookRunner, preParsingHookRunner, onResolutionHookRunner } = require('./hooks')
```
Comment (mcollina) [inline lib/handlers.js:1]: use strict is missing

**Review by mcollina — COMMENTED** [review]
Good work, one last nit.

File: lib/handlers.js (hunk 40 above)
Comment (projectjudge) [reply lib/handlers.js:3]: ah yes - fixed! :)

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you link the new docs pages from the README as well?

**Comment by projectjudge** [comment]
> Could you link the new docs pages from the README as well?

Sure thing!

> Good work! How do the hooks play with subscriptions?

After a bit of experimentation, I think we need to add a few more hook events within `lib/subscription-connection.js`. From my observations with subscriptions so far, I have seen that `preGatewayExecution` is executed when using federated subscriptions. We don't see any other hooks triggered because subscriptions don't go through the normal request route. To fix this, I suggest we add `preParsing`, `preExecution` and `onResolution` (not `preValidation` because we don't do any validation in subscriptions) to `lib/subscription-connection.js` to make sure we capture the full GraphQL lifecycle for subscriptions.

My only concern with this approach is the meaning of context when applied to the hooks. Because it is not strictly associated with a request over the wire, one could argue that we should have subscription specific hooks for this situation instead. I wondered what your thoughts were on this?

Ps, apologies for my slower than normal responses, I'm on holiday this week! :)

**Review by mcollina — APPROVED** [review]
lgtm

**Review by PabloSzx — APPROVED** [review]

**Comment by mcollina** [comment]
> My only concern with this approach is the meaning of context when applied to the hooks. Because it is not strictly associated with a request over the wire, one could argue that we should have subscription specific hooks for this situation instead. I wondered what your thoughts were on this?

It's better to have a subscription specific hook, ideally in a follow up pull request. This is good to ship right now, so I'm shipping it as a few people are waiting on this!

**Comment by projectjudge** [comment]
Yep, makes sense - happy to do a follow up PR for this as you suggested! Thanks for shipping this afternoon! :)

File: docs/hooks.md [hunk 41]
```diff
@@ -0,0 +1,144 @@
...
+  await registerTraceEvent()
+})
+```
+
+### preValidation
+
+By the time the `preValidation` hook triggers, the query string has been parsed into a GraphQL Document AST.
+
+```js
+fastify.addHook('preValidation', async (schema, document, context) => {
```
Comment (SimenB) [inline docs/hooks.md:47]: should this be `fastify.graphql.addHook`? only the first example in this document has that, not the others

Comment (jonnydgreen) [reply docs/hooks.md:47]: You're correct, yep - I'll submit a PR to fix this over lunch!
